components/script/dom/idbrequest.rs

```diff
@@ -0,0 +1,273 @@
+        )
+    }
+
+    pub fn set_source(&self, source: Option<&IDBObjectStore>) {
+        self.source.set(source);
+    }
+
+    pub fn set_ready_state_done(&self) {
+        self.ready_state.set(IDBRequestReadyState::Done);
+    }
+
+    pub fn set_result(&self, result: HandleValue) {
+        self.result.set(result.get());
+    }
+
+    pub fn set_error(&self, error: Error) {
+        // FIXME:(rasviitanen) Support all error types
+        if let Error::Version = error {
+            self.error.set(Some(&DOMException::new(
+                &self.global(),
+                DOMErrorName::VersionError,
+            )));
+        }
+    }
+
+    pub fn set_transaction(&self, transaction: &IDBTransaction) {
+        self.transaction.set(Some(transaction));
+    }
+
+    // https://www.w3.org/TR/IndexedDB-2/#asynchronously-execute-a-request
+    pub fn execute_async(
+        source: &IDBObjectStore,
+        operation: AsyncOperation,
+        request: Option<DomRoot<IDBRequest>>,
+    ) -> Fallible<DomRoot<IDBRequest>> {
+        // Step 1: Let transaction be the transaction associated with source.
+        let transaction = source.transaction().expect("Store has no transaction");
+        let global = transaction.global();
+        let window = global.as_window();
+
+        // Step 2: Assert: transaction is active.
+        if !transaction.is_active() {
+            return Err(Error::TransactionInactive);
+        }
+
+        // Step 3: If request was not given, let request be a new request with source as source.
+        let request = request.unwrap_or({
+            let new_request = IDBRequest::new(window);
+            new_request.set_source(Some(source));
+            new_request.set_transaction(&transaction);
+            new_request
+        });
+
+        // Step 4: Add request to the end of transaction’s request list.
+        transaction.add_request(&request);
+
+        // Step 5: Run the operation, and queue a returning task in parallel
+        // the result will be put into `receiver`
+        let transaction_mode = match transaction.get_mode() {
+            IDBTransactionMode::Readonly => IndexedDBTxnMode::Readonly,
+            IDBTransactionMode::Readwrite => IndexedDBTxnMode::Readwrite,
+            IDBTransactionMode::Versionchange => IndexedDBTxnMode::Versionchange,
+        };
+
+        let (sender, receiver) = ipc::channel(global.time_profiler_chan().clone()).unwrap();
+
+        transaction
+            .global()
+            .resource_threads()
+            .sender()
+            .send(IndexedDBThreadMsg::Async(
+                transaction.get_serial_number(),
+                transaction_mode,
+                sender,
+                operation,
+            ))
+            .unwrap();
+
+        let (task_source, canceller) = window
+            .task_manager()
+            .dom_manipulation_task_source_with_canceller();
+
+        let trusted_request = Trusted::new(&*request);
+        Builder::new()
```

Comment:
For this it could be convenient to use the IPC router, as is done for example at https://github.com/servo/servo/blob/2cf9a00c9983bf23eff23d56321973d36a14f977/components/script/image_listener.rs#L33